cmake/FindLibClang.cmake

```diff
@@ -1,152 +1,167 @@
-FIND_AND_ADD_CLANG_LIB(clangFrontend)
-FIND_AND_ADD_CLANG_LIB(clangSerialization)
-FIND_AND_ADD_CLANG_LIB(clangDriver)
-FIND_AND_ADD_CLANG_LIB(clangCodeGen)
-FIND_AND_ADD_CLANG_LIB(clangParse)
-FIND_AND_ADD_CLANG_LIB(clangSema)
-FIND_AND_ADD_CLANG_LIB(clangAnalysis)
-FIND_AND_ADD_CLANG_LIB(clangAST)
-FIND_AND_ADD_CLANG_LIB(clangEdit)
-FIND_AND_ADD_CLANG_LIB(clangLex)
-FIND_AND_ADD_CLANG_LIB(clangBasic)
-
-list (REMOVE_DUPLICATES LIBCLANG_LIBRARY_DIR)
-
-## -----------------------------------------------------------------------------
-## Actions taken when all components have been found
-
-if (LIBCLANG_INCLUDES AND LIBCLANG_LIBRARY)
-  set (HAVE_LIBCLANG TRUE)
-else (LIBCLANG_INCLUDES AND LIBCLANG_LIBRARY)
-  if (NOT LIBCLANG_FIND_QUIETLY)
-    if (NOT LIBCLANG_INCLUDES)
-      message (STATUS "Unable to find libClang header files!")
-    endif (NOT LIBCLANG_INCLUDES)
-    if (NOT LIBCLANG_LIBRARY)
-      message (STATUS "Unable to find libClang library files!")
-    endif (NOT LIBCLANG_LIBRARY)
-  endif (NOT LIBCLANG_FIND_QUIETLY)
-endif (LIBCLANG_INCLUDES AND LIBCLANG_LIBRARY)
-
-if (HAVE_LIBCLANG)
-  if (NOT LIBCLANG_FIND_QUIETLY)
-    message (STATUS "Found components for libClang")
-    message (STATUS "LIBCLANG_INCLUDES .. = ${LIBCLANG_INCLUDES}")
-    message (STATUS "LIBCLANG_LIBRARIES . = ${LIBCLANG_LIBRARIES}")
-  endif (NOT LIBCLANG_FIND_QUIETLY)
-else (HAVE_LIBCLANG)
-  if (LIBCLANG_FIND_REQUIRED)
-    message (WARNING "Could not find libClang!")
-  endif (LIBCLANG_FIND_REQUIRED)
-endif (HAVE_LIBCLANG)
-
-
-mark_as_advanced (
-    HAVE_LIBCLANG
-    LIBCLANG_INCLUDES
-    LIBCLANG_LIBRARIES
-    LIBCLANG_LIBRARY_DIR
-  )
+if(NOT LIBCLANG_LIBRARY)
+  message(STATUS "LibClang library ... = NOT FOUND")
+else()
+  message(STATUS "LibClang library ... = ${LIBCLANG_LIBRARY}")
+endif()
+
+get_filename_component(LIBCLANG_LIBRARY_DIR ${LIBCLANG_LIBRARY} PATH)
+
+macro(FIND_AND_ADD_CLANG_LIB _libname_)
+  find_library(
+    CLANG_${_libname_}_LIB
+    NAMES ${_libname_}
+    PATHS ${LLVM_LIBRARY_DIR})
+  if(CLANG_${_libname_}_LIB)
+    set(LIBCLANG_LIBRARIES ${LIBCLANG_LIBRARIES} ${CLANG_${_libname_}_LIB})
+    get_filename_component(LIBCLANG_${_libname_}_LIBRARY_DIR
+                           ${LIBCLANG_LIBRARY} PATH)
+    set(LIBCLANG_LIBRARY_DIR ${LIBCLANG_LIBRARY_DIR}
+                             ${LIBCLANG_${_libname_}_LIBRARY_DIR})
+  else()
+    if(LIBCLANG_FIND_VERBOSE)
+      message(
+        WARNING
+          "Clang dependency: Cannot find library '${_libname_}' in paths '${LLVM_LIBRARY_DIR}'."
+      )
+    endif(LIBCLANG_FIND_VERBOSE)
+  endif(CLANG_${_libname_}_LIB)
+endmacro(FIND_AND_ADD_CLANG_LIB)
+
+set(LIBCLANG_LIBRARIES)
+# check if is being linked as shared or not
+if(NOT LLVM_CONFIG_EXECUTABLE)
+  find_program(
+    LLVM_CONFIG_EXECUTABLE
+    NAMES "llvm-config-${LLVM_VERSION_MAJOR}" "llvm-config"
+    DOC "llvm-config executable"
+    PATHS ${LLVM_TOOLS_BINARY_DIR})
+endif(NOT LLVM_CONFIG_EXECUTABLE)
```

Comment:
The comment says we're checking if clang should be linked as shared or not, but no check is done. This check however is needed because if you build LLVM with LLVM_BUILD_LLVM_DYLIB=ON, then

- both static and dynamic libraries will be available for llvm and clang
- llvm-config will return the static libraries by default (at least on macOS)
- FindLibClang.cmake will find libclang-cpp.so which is dynamic

Since both libLLVM.so (through libclang) and the static LLVM libraries are linked together this causes double definitions and in general the built binaries to fail.

You need to run `llvm-config --shared-mode` and check what it returns before searching for dynamic/static clang libraries.

Reply:
I tried this approach, I am testing it with dynamic builds ("shared" ones) but I am failing in being able to test it with a static build. I had found a static build ready [here](https://github.com/savi-lang/llvm-static)I do either get linking errors when using my compiler such as
```libclangFrontend.a(TextDiagnostic.cpp.o): relocation R_X86_64_32S against symbol `_ZTVN5clang14TextDiagnosticE' can not be used when making a PIE object; recompile with -fPIE
``` 
Or I get compiling errors such as:
```/usr/include/wchar.h:35:10: fatal error: 'stddef.h' file not found```
When using the provided compiler.

I'll try to make a static build myself of llvm, it is going to take quite some time I guess

Reply:
Ok, there is a section on the readme about how to build it on MacOS. Let me know if it works as expected!

Reply:
> The libLLVM linking mode should also apply to libclang.

https://github.com/Focshole/libVersioningCompiler/blob/c4c932bed7dd0f156adee5cd80daefe7d33e65f1/cmake/FindLibClang.cmake#L88

I guess this doesn't work.  Do you have any logs with the verbose flags? Which library is the cmake linking against?